### PR TITLE
feat: Add citations implementation for OpenAI, Anthropic, and Google …

### LINKED ISSUE #996: Perplexity AI Citations
### Description

Perplexity AI has a citations field in their response that usually works like this

```python
response_stream = client.chat.completions.create(
    model="sonar-pro",
    messages=messages,
    stream=True,
)

citations = []
for response in response_stream:
    print(response.choices[0].delta.content, end="")
    citations.extend(response.citations)
```

You now have access to a list of citations that you can use reliably.

Is there any way in pydantic_ai to access this?

The perplexity model is not very reliable at producing inline citations. So it would be useful to be able to do this.

### LINKED ISSUE #2128: Support Anthropic tool result citations
### Description

Anthropic has just introduced a new API for citing search result content from custom backends. This is great for RAG applications that return search result data in tools.

### References

https://docs.anthropic.com/en/docs/build-with-claude/search-results

### LINKED ISSUE #2194: Support web search citation output (OpenAI Responses API)
### Description

See [Outputs and citations](https://platform.openai.com/docs/guides/tools-web-search#output-and-citations):

>   Annotations message.content[0].annotations for the cited URLs

### References

_No response_

### LINKED ISSUE #2890: Make Gemini `google_search` tool grounding metadata available in model response
### Description

When the builtin `WebSearchTool` is used with Gemini models, the http response includes the `groundingMetadata` key in `candidates[0]` ([example from their docs](https://ai.google.dev/gemini-api/docs/google-search#understanding_the_grounding_response))

This is the only way to provide proper citations for results found using the builtin tool as far as I'm aware (seemingly no amount of prompting to tell it to include the sources in the text response makes it listen).

Happy to implement this myself if needed but would like some direction on how it should look (if its even something you'd want to add?)

Pushed a simple commit to my fork that was enough to get what I wanted working locally, but I'm doubtful this would be the right approach:

https://github.com/pydantic/pydantic-ai/commit/208ae1256f77618a0bb5f387e15df1bba8bb7565

For more references, here are the types from the google genai lib:
- [CandidateDict](https://github.com/googleapis/python-genai/blob/main/google/genai/types.py#L5243-L5283)
- [GroundingMetadata & GroundingMetadataDict](https://github.com/googleapis/python-genai/blob/main/google/genai/types.py#L4994-L5047)

In my version I then use the result by looking for `BuiltinToolReturnPart` with `tool_name == "grounding_metadata"` to then do what I want with the sources, e.g. something like

```python
def get_grounding_metadata(msgs: list[ModelMessage]) -> GroundingMetadata | None:
    for msg in msgs:
        if not isinstance(msg, ModelResponse):
            continue
        for part in message.parts:
            if isinstance(part, BuiltinToolReturnPart) and part.tool_name == "grounding_metadata":
                if not isinstance(part, GroundingMetadata):
                    raise ValueError('grounding_metadata tool part response not an instance of GroundingMetadata')
                return part
    return None
```

### References

_No response_

### LINKED ISSUE #3126: Support structured citations
### Description

Being implemented in ~~https://github.com/pydantic/pydantic-ai/pull/2657~~ https://github.com/pydantic/pydantic-ai/pull/3729

To do
- [ ] Providers
    - [ ] Sub issues
    - [ ] `OpenRouterModel`

### References

_No response_

### PR DESCRIPTION
**Pull request by certainly-param**
## Summary

This PR implements structured citation support across multiple providers and surfaces them consistently on `TextPart` objects in `ModelResponse`. It builds on the design discussed in [#3126](https://github.com/pydantic/pydantic-ai/issues/3126) and the earlier implementation in [#2657](https://github.com/pydantic/pydantic-ai/pull/2657), rebased on the latest `main`.

## What this changes

- **Core message model**
  - Introduces a generic `Citation` representation (e.g. `URLCitation`, `ToolResultCitation`, `GroundingCitation`).
  - Extends `TextPart` to carry a list of citations, so downstream code can access provider-specific citation data in a uniform way.
  - Wires citations into the existing `ModelMessage` / `ModelResponse` graph.

- **OpenAI (Chat Completions + Responses APIs)**
  - Parses `annotations` from `ChatCompletionMessage` and maps them into `TextPart.citations`.
  - Supports URL-style citations exposed via the OpenAI Responses API, preserving URL, title, and index range metadata.
  - Ensures citations are preserved even when content is split into multiple `TextPart` instances.

- **Anthropic**
  - Maps tool result citations from Anthropic responses into the shared `Citation` representation.
  - Attaches these citations to the relevant `TextPart` instances.

- **Google / Gemini**
  - Parses `grounding_metadata` and `citation_metadata` from Gemini responses.
  - Maps web and file search grounding information into `GroundingCitation` objects.
  - Attaches citations to `TextPart` according to the character / byte ranges provided by Google.

- **Perplexity**
  - Adds a `Perplexity` provider and wires its citation output into the same `Citation` model, so downstream code can consume it alongside the other providers.

- **Docs & examples**
  - Adds API docs for `pydantic_ai.messages` describing the `Citation` types and how they appear on `TextPart.citations`.
  - Adds examples showing how to access and display citations for OpenAI, Anthropic, Google/Gemini, and Perplexity.

- **Tests**
  - Adds unit tests for:
    - OpenAI Chat Completions citations
    - OpenAI Responses API citations
    - Anthropic tool result citations
    - Gemini grounding / web search citations
    - Perplexity citations
    - Edge cases (streaming, message history, OTEL integration, etc.)

## Linked issues / prior work

- Closes #3126  
- Closes #2194  
- Closes #2128  
- Closes #2890  
- Closes #996  
- Supersedes #2657

This PR picks up the work started in #2657, rebases it onto the latest `main`, and extends it to cover the full structured citation story across providers (including Chat Completions).

### CODE REVIEW

File: docs/api/messages.md [hunk 1]
```diff
@@ -17,4 +17,16 @@ graph RL
     ModelResponse("ModelResponse(parts=list[...])") --- ModelMessage("ModelMessage<br>(Union)")
 ```
 
+## Citations
```
Comment (DouweM) [inline docs/api/messages.md:20]: We don't need this here; for API docs purposes this context can be on docstrings.

File: docs/citations/test_coverage.md
Comment (DouweM) [inline docs/citations/test_coverage.md:1]: Unnecessary doc

File: docs/citations/accessing_citations.md [hunk 2]
```diff
@@ -0,0 +1,287 @@
+# Accessing Citations
```
Comment (DouweM) [inline docs/citations/accessing_citations.md:1]: This doc should just be `citations.md`, title `Citations`, and the intro paragraph should introduce the feature and talk about the different types of citations that are supported for which models

Comment (DouweM) [reply docs/citations/accessing_citations.md:1]: In any case, we can come back to docs later and focus on the code first!

File: docs/citations/accessing_citations.md [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+            if isinstance(part, TextPart) and part.citations:
+                for citation in part.citations:
+                    if isinstance(citation, ToolResultCitation):
+                        print(f"Tool: {citation.tool_name}")
+                        print(f"Tool Call ID: {citation.tool_call_id}")
+                        if citation.citation_data:
+                            print(f"Citation Data: {citation.citation_data}")
+```
+
+### GroundingCitation (Google)
```
Comment (DouweM) [inline docs/citations/accessing_citations.md:102]: Do the 3 providers really need their own types? We can't unify to a single format, perhaps with some optional provider-specific fields?

File: docs/examples/citations.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+Citations are references to sources that the model used to generate its response. They typically include URLs, titles, and text ranges indicating which parts of the response are supported by each citation.
+
+## Overview
+
+Pydantic AI supports citations from multiple providers:
+
+- **OpenAI** (Chat Completions and Responses APIs): `URLCitation` with URL, title, and character indices
+- **Anthropic**: `ToolResultCitation` from tool execution results
+- **Google/Gemini**: `GroundingCitation` from grounding metadata
+- **OpenRouter**: Uses OpenAI-compatible citation format
```
Comment (DouweM) [inline docs/examples/citations.md:14]: Note that OpenRouter supports file annotations as well, which are not part of the OpenAI spec:

https://github.com/pydantic/pydantic-ai/blob/d2b08ad12e103e81bd0a9dc0eb7f2a7c3472931d/pydantic_ai_slim/pydantic_ai/models/openrouter.py#L349-L360

File: docs/examples/citations.md [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+- **Perplexity**: Uses OpenAI-compatible citation format
```
Comment (DouweM) [inline docs/examples/citations.md:15]: A user also requested support for Bedrock: https://github.com/pydantic/pydantic-ai/issues/3126#issuecomment-3398189981

File: docs/examples/citations.md [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+For more information, see the [Citations Overview](../citations/overview.md) and [Accessing Citations](../citations/accessing_citations.md) guides.
```
Comment (DouweM) [inline docs/examples/citations.md:17]: Overview doc doesnt' exist

File: docs/examples/citations.md
Comment (DouweM) [inline docs/examples/citations.md:1]: I don't think this feature needs examples

File: pydantic_ai_slim/pydantic_ai/__init__.py [hunk 7]
```diff
@@ -44,7 +44,7 @@
     BinaryImage,
     BuiltinToolCallPart,
     BuiltinToolReturnPart,
-    CachePoint,
```
Comment (DouweM) [inline pydantic_ai_slim/pydantic_ai/__init__.py:47]: `CachePoint` was removed accidentally!

File: pydantic_ai_slim/pydantic_ai/messages.py [hunk 8]
```diff
@@ -1021,6 +1021,103 @@ def user_text_prompt(cls, user_prompt: str, *, instructions: str | None = None)
     __repr__ = _utils.dataclasses_no_defaults_repr
 
 
+@dataclass(repr=False)
+class URLCitation:
+    """A citation with a URL, used by OpenAI and similar providers.
```
Comment (DouweM) [inline pydantic_ai_slim/pydantic_ai/messages.py:1026]: Can we also map Anthropic web search citations (https://github.com/pydantic/pydantic-ai/blob/d2b08ad12e103e81bd0a9dc0eb7f2a7c3472931d/tests/models/cassettes/test_anthropic/test_anthropic_web_search_tool.yaml#L142-L149) and Google web search annotations to the same data model?

Comment (DouweM) [reply pydantic_ai_slim/pydantic_ai/messages.py:1026]: Ultimately the goal of native citations support is to give users 1 way to deal with all citations from all providers; not to have them write provider-specific code. So we should try our hardest to unify the data model.s

File: pydantic_ai_slim/pydantic_ai/messages.py [hunk 9]
```diff
@@ -1021,6 +1021,103 @@ def user_text_prompt(cls, user_prompt: str, *, instructions: str | None = None)
...
+
+    _: KW_ONLY
+
+    tool_call_id: str | None = None
+    """ID of the tool call that generated this citation."""
+
+    citation_data: dict[str, Any] | None = None
+    """Extra citation data from the tool result.
+
+    Structure varies by provider.
```
Comment (DouweM) [inline pydantic_ai_slim/pydantic_ai/messages.py:1116]: Why if this class is Anthropic-specific anyway?

File: pydantic_ai_slim/pydantic_ai/messages.py [hunk 10]
```diff
@@ -1021,6 +1021,103 @@ def user_text_prompt(cls, user_prompt: str, *, instructions: str | None = None)
...
+
+    _: KW_ONLY
+
+    grounding_metadata: dict[str, Any] | None = None
+    """Grounding metadata from the response.
+
+    Has info about sources used for grounding.
+    """
+
+    citation_metadata: dict[str, Any] | None = None
```
Comment (DouweM) [inline pydantic_ai_slim/pydantic_ai/messages.py:1137]: As mentioned, I'd like to parse this into a unified data model equivalent to what we get for OpenAI and Anthropic (web search) results.

I think it may be possible for the `TextPart.citations` list to only have a single type `Citation`, and then for a `Citation` to have a `source` that could be `Url` or `File` or `ToolResult` or something else.

File: pydantic_ai_slim/pydantic_ai/models/__init__.py [hunk 11]
```diff
@@ -1019,18 +1019,20 @@ def infer_model(  # noqa: C901
...
-        'ollama',
+        'openai',
+        'openai-chat',
+        'openrouter',
+        'perplexity',
         'together',
         'vercel',
-        'litellm',
-        'nebius',
-        'ovhcloud',
```
Comment (DouweM) [inline pydantic_ai_slim/pydantic_ai/models/__init__.py:1033]: We lost a bunch of stuff here! Please review the entire PR line by line before passing it to me for review :)

File: tests/test_citation_message_history.py [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+        model_name='test',
+        usage=usage.RequestUsage(input_tokens=10, output_tokens=5),
+    )
+    messages = [response]
+
+    # Serialize and deserialize
+    json_bytes = ModelMessagesTypeAdapter.dump_json(messages)
+    deserialized = ModelMessagesTypeAdapter.validate_python(ModelMessagesTypeAdapter.validate_json(json_bytes))
+
+    assert len(deserialized) == 1
```
Comment (DouweM) [inline tests/test_citation_message_history.py:27]: Please use `snapshot`s in tests wherever possible, so we can see the entire resulting structure rather than testing field by field (and possible missing some)

**Review by DouweM — CHANGES_REQUESTED** [review]
@certainly-param Thanks for working on this! Before we continue coding though, I'd like to spend a bit more time working out the design of our unified interface, as the goal of this feature is to let users deal with citations from all providers in one consistent way, instead of having to know that OpenAI returns X, Anthropic returns Y, Google returns Z etc.

So before we get to coding, we need to know all of the diverse data representations we're dealing with, and then come up with something that covers them all (with mostly fields we can fill in for each provider, and then some optional or `provider_details` fields for extra provider-specific stuff). 

So as a first step, can you please gather the type definitions for all the annotations/citations/grounding data the various providers support and post it in the issue, with a proposal of how we could represent that in a unified way? Ideally we'd also have samples of real data, which you can likely find in existing Pydantic AI cassettes (e.g. look for `cited_text` to find Anthropic data; `groundingMetadata` to find Google data, etc). 

I think it may be possible for the `TextPart.citations` list to only have a single type `Citation`, and then for a `Citation` to have a `source` field that could be `Url` or `File` or `ToolResult` or something else that's revealed by that research into all the shapes the provider data can take.

**Comment by hayescode** [comment]
Will Azure's version of OpenAI's responses API web_search be included here? See #3698

Microsoft doesn't list the URL sources in the web_search events and only adds then to annotations in the final response. I have no idea why Microsoft doesn't match OpenAI's perfectly good responses API, it is maddening.

**Comment by DouweM** [comment]
@hayescode Yep, fortunately Azure is using the same annotations/citations field that OpenAI uses, and those are exactly the fields we're looking to support here.

OpenAI doesn't include web search sources by default either, requiring to be set https://ai.pydantic.dev/api/models/openai/#pydantic_ai.models.openai.OpenAIResponsesModelSettings.openai_include_web_search_sources. Does Azure not support that / something equivalent?

**Comment by hayescode** [comment]
> @hayescode Yep, fortunately Azure is using the same annotations/citations field that OpenAI uses, and those are exactly the fields we're looking to support here.
> 
> OpenAI doesn't include web search sources by default either, requiring to be set https://ai.pydantic.dev/api/models/openai/#pydantic_ai.models.openai.OpenAIResponsesModelSettings.openai_include_web_search_sources. Does Azure not support that / something equivalent?

@DouweM Azure OpenAI does match the shape of the Responses API, but they don't populate the "source" or anything and instead add the sources to the final annotations. This is because Azure doesn't use OpenAI's web search, they ruined the best web_search tool by making their own **Bing** version. Here's the documentation: https://learn.microsoft.com/en-us/azure/ai-foundry/openai/how-to/web-search?view=foundry-classic

```json
  "response": {
    "id": "resp_035bd9a24ffa66a00069441485f7a48190aa254c61ac5a619b",
    "created_at": 1766069381.0,
    "error": null,
    "incomplete_details": null,
    "instructions": null,
    "metadata": {},
    "model": "gpt-5.2",
    "object": "response",
    "output": [
      {
        "id": "ws_035bd9a24ffa66a000694414869efc81909ea11f148f1d994a",
        "action": {
          "query": null,
          "type": "search",
          "sources": null
        },
        "status": "completed",
        "type": "web_search_call"
      },
      {
        "id": "ws_035bd9a24ffa66a0006944148724148190b8f665fff7e18d83",
        "action": {
          "query": "current weather St. Louis MO right now",
          "type": "search",
          "sources": null
        },
        "status": "completed",
        "type": "web_search_call"
      },
      {
        "id": "msg_035bd9a24ffa66a0006944148e09d88190b1330eadedcdf2c6",
        "content": [
          {
            "annotations": [
              {
                "end_index": 659,
                "start_index": 541,
                "title": "St. Louis, MO Current Weather | AccuWeather",
                "type": "url_citation",
                "url": "https://www.accuweather.com/en/us/st-louis/63102/current-weather/349084"
              },
              {
                "end_index": 789,
                "start_index": 659,
                "title": "Saint Louis, MO Current Weather - The Weather Network",
                "type": "url_citation",
                "url": "https://www.theweathernetwork.com/en/city/us/missouri/saint-louis/current"
              },
              {
                "end_index": 873,
                "start_index": 789,
                "title": "Weather today - St. Louis, MO",
                "type": "url_citation",
                "url": "https://www.weather-us.com/en/missouri-usa/st-louis"
              }
            ],
            "text": "I can’t reliably give “current” weather from these results because they’re inconsistent and appear to be from different dates/times. For example, AccuWeather shows **54°F and sunny at 1:05 PM** , while The Weather Network shows **-13°C (about 9°F) and clear** updated minutes ago , and Weather Channel’s hourly view is stamped for **Monday, Dec 15** .\n\nTell me your **ZIP code (or nearest neighborhood)** and whether you want it **right now** or **today’s forecast**, and I’ll pull the most consistent current conditions for your exact spot.[St. Louis, MO Current Weather | AccuWeather](https://www.accuweather.com/en/us/st-louis/63102/current-weather/349084)[Saint Louis, MO Current Weather - The Weather Network](https://www.theweathernetwork.com/en/city/us/missouri/saint-louis/current)[Weather today - St. Louis, MO](https://www.weather-us.com/en/missouri-usa/st-louis)",
            "type": "output_text",
            "logprobs": []
          }
        ],
        "role": "assistant",
        "status": "completed",
        "type": "message"
      }
    ],
    "parallel_tool_calls": true,
    "temperature": 1.0,
    "tool_choice": {
      "type": "web_search_preview"
    },
    "tools": [
      {
        "type": "web_search",
        "filters": null,
        "search_context_size": "medium",
        "user_location": {
          "city": "St Louis",
          "country": "US",
          "region": "MO",
          "timezone": null,
          "type": "approximate"
        }
      }
    ],
    "top_p": 0.98,
    "background": false,
    "conversation": null,
    "max_output_tokens": null,
    "max_tool_calls": null,
    "previous_response_id": null,
    "prompt": null,
    "prompt_cache_key": null,
    "prompt_cache_retention": null,
    "reasoning": {
      "effort": "none",
      "generate_summary": null,
      "summary": null
    },
    "safety_identifier": null,
    "service_tier": "default",
    "status": "completed",
    "text": {
      "format": {
        "type": "text"
      },
      "verbosity": "medium"
    },
    "top_logprobs": 0,
    "truncation": "disabled",
    "usage": {
      "input_tokens": 6232,
      "input_tokens_details": {
        "cached_tokens": 0
      },
      "output_tokens": 236,
      "output_tokens_details": {
        "reasoning_tokens": 65
      },
      "total_tokens": 6468
    },
    "user": null,
    "content_filters": null,
    "store": true
  },
  "sequence_number": 34,
  "type": "response.completed"
```

**Comment by hayescode** [comment]
> This PR is stale, and will be closed in 3 days if no reply is received.

Keep

**Comment by certainly-param** [comment]
> @certainly-param Thanks for working on this! Before we continue coding though, I'd like to spend a bit more time working out the design of our unified interface, as the goal of this feature is to let users deal with citations from all providers in one consistent way, instead of having to know that OpenAI returns X, Anthropic returns Y, Google returns Z etc.
> 
> So before we get to coding, we need to know all of the diverse data representations we're dealing with, and then come up with something that covers them all (with mostly fields we can fill in for each provider, and then some optional or `provider_details` fields for extra provider-specific stuff).
> 
> So as a first step, can you please gather the type definitions for all the annotations/citations/grounding data the various providers support and post it in the issue, with a proposal of how we could represent that in a unified way? Ideally we'd also have samples of real data, which you can likely find in existing Pydantic AI cassettes (e.g. look for `cited_text` to find Anthropic data; `groundingMetadata` to find Google data, etc).
> 
> I think it may be possible for the `TextPart.citations` list to only have a single type `Citation`, and then for a `Citation` to have a `source` field that could be `Url` or `File` or `ToolResult` or something else that's revealed by that research into all the shapes the provider data can take.

Thanks for this review! I've created issue #3885 with research on provider-specific citation/annotation/grounding types and sample data from cassettes

File: docs/api/messages.md (hunk 1 above)
Comment (certainly-param) [reply docs/api/messages.md:20]: Done. I've removed the Citations section from docs/api/messages.md. The citation context is now only in the docstrings (as it should be for API docs).

File: docs/citations/test_coverage.md
Comment (certainly-param) [reply docs/citations/test_coverage.md:1]: Done. I've deleted docs/citations/test_coverage.md.

File: docs/citations/accessing_citations.md (hunk 2 above)
Comment (certainly-param) [reply docs/citations/accessing_citations.md:1]: Done. I've renamed docs/citations/accessing_citations.md to docs/citations/citations.md, changed the title to "Citations", and updated the intro paragraph to introduce the feature and explain the different citation types supported by each provider/model.

File: docs/citations/accessing_citations.md (hunk 3 above)
Comment (certainly-param) [reply docs/citations/accessing_citations.md:102]: This design question is covered in the research and unified design proposal in issue #3885. The proposal suggests a single citation type with a discriminated union source field (UrlSource, FileSource, ToolResultSource) to unify all provider-specific citation formats.

File: docs/examples/citations.md (hunk 4 above)
Comment (certainly-param) [reply docs/examples/citations.md:14]: Noted. The information about OpenRouter file annotations is already documented in the research in issue #3885.
I've deleted docs/examples/citations.md per your earlier comment, this feature doesn't need examples.

File: docs/examples/citations.md (hunk 5 above)
Comment (certainly-param) [reply docs/examples/citations.md:15]: Noted. The Bedrock citation support request from issue #3126 is already included in the research documented in issue #3885. The research notes that Bedrock citation support needs to be investigated.

File: docs/examples/citations.md (hunk 6 above)
Comment (certainly-param) [reply docs/examples/citations.md:17]: Noted. The file docs/examples/citations.md has been deleted per your earlier comment, so the broken reference to overview.md is no longer an issue.

File: docs/examples/citations.md
Comment (certainly-param) [reply docs/examples/citations.md:1]: Done. Already Removed.

File: pydantic_ai_slim/pydantic_ai/__init__.py (hunk 7 above)
Comment (certainly-param) [reply pydantic_ai_slim/pydantic_ai/__init__.py:47]: Done. I've restored CachePoint in pydantic_ai/__init__.py - added it back to the imports from .messages and to the __all__ tuple.

File: pydantic_ai_slim/pydantic_ai/messages.py (hunk 8 above)
Comment (certainly-param) [reply pydantic_ai_slim/pydantic_ai/messages.py:1026]: This design question is addressed in the unified citation design proposal in issue #3885. The proposal is a starting point for discussion, and there might be better approaches. I'd appreciate discussion!

File: pydantic_ai_slim/pydantic_ai/messages.py (hunk 9 above)
Comment (certainly-param) [reply pydantic_ai_slim/pydantic_ai/messages.py:1116]: Noted. This inconsistency can be addressed as part of #3885.

File: pydantic_ai_slim/pydantic_ai/messages.py (hunk 10 above)
Comment (certainly-param) [reply pydantic_ai_slim/pydantic_ai/messages.py:1137]: Can be discussed in issue #3885

File: pydantic_ai_slim/pydantic_ai/models/__init__.py (hunk 11 above)
Comment (certainly-param) [reply pydantic_ai_slim/pydantic_ai/models/__init__.py:1033]: Sorry about that. I've restored nebius and ovhcloud in the infer_model function's provider list.

File: tests/test_citation_message_history.py (hunk 12 above)
Comment (certainly-param) [reply tests/test_citation_message_history.py:27]: Done. I've updated the citation serialization tests in test_citation_message_history.py to use snapshots instead of field-by-field assertions. The tests now use snapshot(None) which will be filled in by inline-snapshot when running the tests, allowing us to see the entire resulting structure rather than testing individual fields.

**Comment by prassanna-ravishankar** [comment]
bump! been looking forward to citations since october.

File: docs/api/messages.md [hunk 13]
```diff
@@ -17,4 +17,16 @@ graph RL
...
+
+[`TextPart`][pydantic_ai.messages.TextPart] objects can include citations that reference sources used by the model. Citations are stored in the `citations` field and can be one of three types:
```
Comment (hayescode) [inline docs/api/messages.md:22]: @certainly-param @DouweM OpenAI emits `container_file_citation` in addition to `url_citation`. This happens when OpenAI's built-in code interpreter tool creates files. Will this be handled? Right now these are unavailable in PydanticAI and only if the code interpreter produces an image is it automatically made into a FilePart (ideally auto-downloading files should be for all file mimes not just images 🤞)

https://platform.openai.com/docs/guides/tools-code-interpreter#work-with-files

Edit: Looks like only `url_citation` is handled. The only other type i'm aware of is `container_file_citation`.

https://github.com/certainly-param/pydantic-ai/blob/4c573470ce826d1d100f34e355a5c5cdd47398c2/pydantic_ai_slim/pydantic_ai/models/openai.py#L621-L622

Comment (DouweM) [reply docs/api/messages.md:22]: > (ideally auto-downloading files should be for all file mimes not just images 🤞)

@hayescode Issue/PR more than welcome!

**Comment by DouweM** [comment]
Marking as draft while we discuss in https://github.com/pydantic/pydantic-ai/issues/3885#issuecomment-3734855382.

File: docs/api/messages.md (hunk 13 above)
Comment (hayescode) [reply docs/api/messages.md:22]: @DouweM i'll take a swing at it. I actually found several more here. https://github.com/openai/openai-python/blob/main/src/openai/types/responses/response_output_text.py

@certainly-param if i can get these added i'll open a PR to this fork to include in this PR.

Comment (hayescode) [reply docs/api/messages.md:22]: Issue: #4017

Comment (hayescode) [reply docs/api/messages.md:22]: @certainly-param I added ContainerFileCitation support to your branch and opened a PR if you can kindly please review merge :)

https://github.com/certainly-param/pydantic-ai/pull/2

**Comment by DouweM** [comment]
@certainly-param Are you still interested in working on this?

**Comment by certainly-param** [comment]
> @certainly-param Are you still interested in working on this?

Yes, I am. Was caught in something. I'll be able to work on this starting next week.

**Comment by certainly-param** [comment]
> @certainly-param Are you still interested in working on this?

@DouweM Do you want to have discussions about this PR somewhere else? I see that we don't have a discussion section here.

**Comment by corytomlinson** [comment]
Hi guys, wondering where things are with this? Would be great to see it cross the finish line.

**Comment by DouweM** [comment]
@certainly-param Sorry for the delay here; if you can [join Slack](https://logfire.pydantic.dev/docs/join-slack/) we can sync more efficiently there in text or on a video call!

**Comment by mattbrandman** [comment]
@certainly-param  are you still working on this / have you aligned with the pydantic team in slack? Feature is definitely of interest to our team

**Comment by certainly-param** [comment]
> @certainly-param are you still working on this / have you aligned with the pydantic team in slack? Feature is definitely of interest to our team

Hi, sorry for the delay. Yes, I have just joined slack. I definitely want to work on this!
